### ISSUE TITLE
forward log loss after client reboot ?

### ISSUE THREAD

**Issue by yueguifang** [rsyslog/rsyslog-pkg-ubuntu#97]
Forwarding log to a remote server fail, after I reboot my system, this log cannot forward to remote server again. Does rsyslog support this behavior？

**Comment by davidelang** [comment]
yes, what does rsyslog report about the failure? do you have a queue for the 
send to the remote system so that it can continue to log locally?

David Lang

**Comment by yueguifang** [comment]
Send to a remote system so it can continue to log in locally?
My client is a removable device, often shut down, and often in the absence of a network. I hope that when there is no network, the log can be cached locally and the log can be forwarded to the remote server after restarting. Below is me Configuration, please help to see why it ca n’t be stored locally。


$ActionQueueFileName fwdRule1 # unique name prefix for spool files
$ActionQueueMaxFileSize 10M
$ActionQueueMaxDiskSpace 1g   # 1gb space limit (use as much as possible)
$ActionQueueSaveOnShutdown on # save messages to disk on shutdown
$ActionQueueType Disk # run asynchronously
$ActionResumeRetryCount -1    # infinite retries if host is down
$ActionQueueDiscardSeverity 8
$ActionQueueTimeoutEnqueue 3000
$ActionQueueDequeueBatchSize 500
$MainMsgQueueSaveOnShutdown on
#$DefaultNetstreamDriver gtls

#$DefaultNetstreamDriverCAFile /Surgery/Config/Certificate/Trust/trustCert.pem
#$DefaultNetstreamDriverCertFile /Surgery/Config/Certificate/Host/Pem/cert.pem
#$DefaultNetstreamDriverKeyFile /Surgery/Config/Certificate/Host/Pem/rsa.key

#$ActionSendStreamDriverAuthMode x509/certvalid
#$ActionSendStreamDriverMode 1

# remote host is: name/ip:port, e.g. 192.168.0.1:514, port optional
#*.* @@remote-host:514
$template tlsFormat,"GE %timestamp% %hostname% %syslogtag% %msg%"
# ### remote server ###
authpriv.* @@10.190.63.240:514;tlsFormat
#*.info;mail.none;authpriv.none @@10.190.63.240:514;tlsFormat
mail.* @@10.190.63.240:514;tlsFormat
uucp,news.crit @@10.190.63.240:514;tlsFormat

**Comment by yueguifang** [comment]
supplement： I am not recieve the failure message, and

**Comment by davidelang** [comment]
Ok, that is a bit of a different situation, instead of just needing to handle a 
temporary outage, the normal situation is to not have the remote receiver 
available.

first off, looking at your config, you only have a queue on the authpriv action, 
not on any of the others. This is not obvious because you are using the obsolete 
syntax that sets all the queue parameters with $actionqueue* lines. These only 
apply to the next action, not to all following actions. As a result, when the 
mail logs can't be sent, all log processing blocks.

I am curious why you are sending your logs in a format that Rsyslog is not going 
to parse properly on the receiving side (your tlsformat), which also throws away 
the facility/severity information that the standard syslog format provides.

Probably the best thing for you to do is to create a ruleset and define a queue 
on that ruleset. This way you only have one queue with one copy of the logs 
rather than a separate queue/coopy for each action.

Also, setting the queue type to disk will slow your log processing by 100x-1000x 
and is not recommended, use FixedArray or LinkedList and rsyslog will still 
write things out to disk when the in-memory array is full (or when you shutdown)

you also need to make sure that your system shutdown process gives rsyslog time 
to write out all the messages in the queue. By default, systemd only gives it a 
short time before issuing a kill -9.

David Lang

**Comment by davidelang** [comment]
Assuming that you are using a current version of rsyslog (released within the 
last 6-9 months), you can start it with the -o command line flag and have it 
output it's interpretation of the config that you have provided it. I think you 
would be very surprised at what it shows. (and this is the reason the syntax you 
are using is labeled 'obsolete' and discouraged)

David Lang

**Comment by yueguifang** [comment]
Thank you very much for your kindly reply.
you say I use the the obsolete syntax to set the queue parameters, and suggest me to create a ruleset and define a queue on that ruleset. Can you give me an example for reference.
